# zAMExplorer
zAMPExplorer is an interactive Shiny application designed to facilitate the downstream analysis of outputs generated by the [zAMP pipeline](https://zamp.readthedocs.io/en/latest/) . zAMP itself is a comprehensive DADA2-based bioinformatics pipeline for processing 16S rRNA gene amplicon metagenomic sequences, offering convenient, reproducible, and scalable analysis from raw sequencing reads to taxonomic profiles. The output of zAMP is a phyloseq object, which serves as the input for zAMPExplorer.

zAMPExplorer enables users to perform a wide range of microbiota and statistical analyses, including compositional barplot, relative abundance heatmap, community diversity (alpha diversity), community similarity through unsupervised (NMDS/PCoA) and supervised (RDA) ordinations, and community typing (or clustering) of microbial profiles using Dirichlet Multinomial Mixtures (DMM). All of these analyses are made accessible through an intuitive graphical interface, bridging the gap between complex command-line bioinformatics processing and user-friendly data exploration.


## Prerequisites and installation

### Prerequisites

- **Operating system**: Windows, macOS, or Linux
- **R**: Version 4.0 or later
- **RStudio**: Recommended for running the Shiny app
- **zAMP**: zAMPExplorer is designed to work with output generated from the zAMP pipeline.

### Installation

1. **Clone the repository**:
   ```bash
   git clone https://github.com/metagenlab/zAMExplorer.git
   ```
2. **Navigate to the project directory**:
   ```bash
   cd zAMExplorer
   ```
3. **Install dependencies**:
   Run the provided `Dependencies.R` script to install all necessary R packages:
   ```r
   source("Dependencies.R")
   ```

4. **Launch the app**:
   After installation, launch zAMPExplorer with:
   ```r
   shiny::runApp("path_to_your_app")
   ```

